### PR TITLE
Container placeholder to facilitate full serialization to and from disk.

### PR DESCRIPTION
**Pull request by brandon-northcutt**
This is a first stab at a wrapper container to support full serialization to disk (#253) using the many various protos. Note: there isn't much meat here, more an approach with various steps outlined. Does this seem like the right approach @SirVer, @wohe, @damonkohler, @macmason?

---
This change is [Reviewable](https://reviewable.io/reviews/googlecartographer/cartographer/335)

### CODE REVIEW

**Comment by brandon-northcutt** [comment]
I signed it!

File: cartographer/io/map_io.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_IO_MAPPING_IO_H_
+#define CARTOGRAPHER_IO_MAPPING_IO_H_
+
+#include "cartographer/mapping/map_builder.h"
+#include "cartographer/io/file_writer.h"//Re-used filestream wrapper.
```
Comment (macmason) [inline cartographer/io/map_io.h]: Extraneous comment here.

File: cartographer/io/map_io.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace cartographer {
+namespace io {
+
+// A class to serialize and deserialize the Cartographer state.
+class MapIO {
+public:
+  MapIO(cartographer::mapping::MapBuilder & map_builder);
```
Comment (macmason) [inline cartographer/io/map_io.h]: GSG requires that `map_builder` either be passed const reference (in the case when it's not modified) or by pointer (in the case where it's modified). 

Also, no space before '&', '*', etc: type modifiers cuddle left in Cartographer code.

File: cartographer/io/map_io.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  MapIO(const MapIO &) = delete;
+  MapIO &operator=(const MapIO &) = delete;
+  //TODO(brandon-northcutt) Write the full Cartographer state to disk.
```
Comment (macmason) [inline cartographer/io/map_io.h]: Space after // in comments (here and lots of other places).

File: cartographer/io/map_io.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  void write(const std::string file_name);
```
Comment (macmason) [inline cartographer/io/map_io.h]: GSG specifies that "non-trivial" functions `BeNamedLikeThis`, so you want `Write`.

It should also return a `bool` signifying success, and (depending on what the Googlers say) maybe also set the warn_unused_result attribute.

File: cartographer/io/map_io.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  //TODO(brandon-northcutt) Read the full cartographer state from disk.
+  void read(const std::string file_name);
```
Comment (macmason) [inline cartographer/io/map_io.h]: Ditto above.

File: cartographer/io/proto/map_io.proto [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto2";
+
+package cartographer.mapping.proto;
```
Comment (macmason) [inline cartographer/io/proto/map_io.proto]: Perhaps `cartographer.io.proto`?

File: cartographer/io/map_io.cc [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/io/map_io.h"
+
+namespace cartographer {
+namespace io {
+
+//constructor takes a pointer to the map_builder
```
Comment (macmason) [inline cartographer/io/map_io.cc]: Comment is wrong.

Comment (brandon-northcutt) [inline cartographer/io/map_io.cc]: Done.

**Comment by brandon-northcutt** [comment]
It looks like the Travis CI build fails with the error:

```bash
cartographer/io/proto/map_io.proto:55:3: Expected "required", "optional", or "repeated".
```
This is on the line defining a `oneof` protobuf message and I get no such error message compiling the file using the current version of protobuf. I think that `oneof` is supported by [proto2](https://developers.google.com/protocol-buffers/docs/proto#oneof). Perhaps the protobuf compiler in Travis CI is due for an update? I'm also fine with using another method that does not require `oneof`  [(#253)](https://github.com/googlecartographer/cartographer/issues/253#issue-225918899).

File: cartographer/io/proto/map_io.proto [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto2";
+
+package cartographer.io.proto;
+
+import "cartographer/common/proto/ceres_solver_options.proto";
```
Comment (wohe) [inline cartographer/io/proto/map_io.proto]: Let's only import what is used.

File: cartographer/io/map_io.cc [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+#include "cartographer/io/map_io.h"
+
+namespace cartographer {
+namespace io {
+
+//constructor takes a const reference to the map_builder
+cartographer::io::MapIO::MapIO(const cartographer::mapping::MapBuilder
+    &map_builder){}
+
+bool MapIO::Write(const string file_name){
```
Comment (wohe) [inline cartographer/io/map_io.cc]: Let's make the serialization part of `MapBuilder`, similar to how there is `SparsePoseGraph::ToProto()`, maybe like this:
```
bool MapBuilder::SerializeState(io::ProtoStreamWriter* writer);
```
There is no need to keep additional state across calls and by putting it at the `MapBuilder` we do not have to expose implementation details.

File: cartographer/io/map_io.cc [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/io/map_io.h"
+
+namespace cartographer {
+namespace io {
+
+//constructor takes a const reference to the map_builder
```
Comment (wohe) [inline cartographer/io/map_io.cc]: Please do not add comments repeating what the code does.

File: cartographer/io/map_io.cc [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+cartographer::io::MapIO::MapIO(const cartographer::mapping::MapBuilder
+    &map_builder){}
+
+bool MapIO::Write(const string file_name){
+  // TODO(brandon-northcutt) Pause optimization so that submaps, trajectories,
+  //  and constraints are immutable for serialization of state.
+
+  // TODO(brandon-northcutt) Get a file handle to send protobufs to.
+
+  // TODO(brandon-northcutt) Send map_builder_options protobuf to file first, so
```
Comment (wohe) [inline cartographer/io/map_io.cc]: I think we should let the deserialization code use a different configuration if necessary.

File: cartographer/io/map_io.cc [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  // TODO(brandon-northcutt) Iterate trajectories { serialize trajectory builder
+  //  options, serialize sensor configuration proto, serialize all nodes in this
+  //  trajectory.
+  
+  // TODO(brandon-northcutt) Serialize trajectory connectivity.
+
+  // TODO(brandon-northcutt) Iterate submaps(serialize submap options, serialize
+  //  probability grids.
+  
+  // TODO(brandon-northcutt) Serialize the sparse pose graph,  (Should the SPG
```
Comment (wohe) [inline cartographer/io/map_io.cc]: I'd suggest serializing this first. Then the big picture is clear after looking at the first proto message in the stream.

**Pull request closed without merging**